File: app/src/main/java/se/curtrune/lucy/viewmodel/ItemSessionViewModel.java
package se.curtrune.lucy.viewmodel;

import static se.curtrune.lucy.util.Logger.log;

import android.content.Context;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import java.time.Duration;
import java.time.LocalTime;
import java.util.ArrayList;
import java.util.List;

import se.curtrune.lucy.R;
import se.curtrune.lucy.classes.Item;
import se.curtrune.lucy.classes.Mental;
import se.curtrune.lucy.classes.Repeat;
import se.curtrune.lucy.item_settings.CheckBoxSetting;
import se.curtrune.lucy.item_settings.ItemSetting;
import se.curtrune.lucy.item_settings.KeyValueSetting;
import se.curtrune.lucy.screens.util.Converter;
import se.curtrune.lucy.util.Kronos;
import se.curtrune.lucy.persist.ItemsWorker;
import se.curtrune.lucy.workers.NotificationsWorker;

public class ItemSessionViewModel extends ViewModel {
    private MutableLiveData<Item> mutableItem;
    private ItemSetting currentItemSetting;
    private final MutableLiveData<Kronos.State> mutableTimerState = new MutableLiveData<>();
    private final MutableLiveData<Long> mutableDuration = new MutableLiveData<>();
    private final MutableLiveData<String> mutableError = new MutableLiveData<>();
    private final MutableLiveData<Item> mutableCurrentItem = new MutableLiveData<>();
    private static final int MENTAL_OFFSET = 5;
    public static boolean VERBOSE = false;
    private Item currentItem;
    private Item currentTimerItem;
    private Kronos kronos;
    public void addTags(String tags, Context context) {
        currentItem.setTags(tags);
        update(currentItem, context);
    }
    public void cancelNotification(Context context) {
        log("ItemSessionViewModel.cancelNotification(Context)");
        NotificationsWorker.cancelNotification(currentItem, context);
        currentItem.setNotification(null);
        update(currentItem, context);
    }
    public void cancelTimer(){
        log("ItemSessionViewModel.cancelTimer()");
        if( kronos == null){
            return;
        }
        kronos.reset();
        mutableDuration.setValue(0L);
        mutableTimerState.setValue(Kronos.State.STOPPED);
    }
    public void init(Item item, Context context){
        log("ItemSessionViewModel.init(Item)");
        assert  item != null;
        this.currentItem = item;
        if(currentItem.getRepeatID() > 0){
            Repeat repeat = ItemsWorker.selectRepeat(currentItem.getRepeatID(), context);
            currentItem.setRepeat( repeat);
        }
        this.mutableCurrentItem.setValue(currentItem);
        this.mutableDuration.setValue(currentItem.getDuration());

    }
    public LiveData<Item> getCurrentItem(){
        return mutableCurrentItem;
    }
    public LiveData<Long> getDuration(){
        return mutableDuration;
    }
    public LiveData<String> getError(){
        return mutableError;
    }
    public int getStress() {
        return currentItem.getStress();
    }
    public int getAnxiety(){
        return currentItem.getAnxiety();
    }
    public int getMood(){
        return currentItem.getMood();
    }
    public int getEnergy(){
        return currentItem.getEnergy() ;
    }
    public List<ItemSetting> getItemSettings(Item item, Context context) {
        assert item != null;
        log("ItemEditorViewModel.getItemSettings(Item)");
        List<ItemSetting> settings = new ArrayList<>();
        KeyValueSetting dateSetting = new KeyValueSetting(context.getString(R.string.date), item.getTargetDate().toString(), ItemSetting.Key.DATE);
        KeyValueSetting timeSetting = new KeyValueSetting(context.getString(R.string.time),item.getTargetTime().toString(), ItemSetting.Key.TIME);
        String repeatValue = "";
        if( item.hasPeriod()){
                repeatValue = item.getPeriod().toString();
        }
        KeyValueSetting repeatSetting = new KeyValueSetting(context.getString(R.string.repeat), repeatValue, ItemSetting.Key.REPEAT);
        KeyValueSetting colourSetting = new KeyValueSetting(context.getString(R.string.color), String.valueOf(item.getColor()), ItemSetting.Key.COLOR);
        CheckBoxSetting templateSetting = new CheckBoxSetting(context.getString(R.string.is_template), item.isTemplate(), ItemSetting.Key.TEMPLATE);
        CheckBoxSetting prioritizedSetting = new CheckBoxSetting( context.getString(R.string.is_prioritized), item.isPrioritized(), ItemSetting.Key.PRIORITIZED);
        KeyValueSetting durationSetting = new KeyValueSetting(context.getString(R.string.duration), Converter.formatSecondsWithHours(item.getDuration()),ItemSetting.Key.DURATION);
        String notificationValue = "";

        if( item.hasNotification()){
            notificationValue = item.getNotification().toString();
        }
        KeyValueSetting notificationSetting = new KeyValueSetting("notification", notificationValue, ItemSetting.Key.NOTIFICATION);
        KeyValueSetting tagsSetting = new KeyValueSetting(context.getString(R.string.tags), item.getTags(), ItemSetting.Key.TAGS);
        CheckBoxSetting calenderSetting = new CheckBoxSetting(context.getString(R.string.is_calender), item.isCalenderItem(), ItemSetting.Key.IS_CALENDAR_ITEM);
        KeyValueSetting categorySetting = new KeyValueSetting(context.getString(R.string.category), item.getCategory(), ItemSetting.Key.CATEGORY);

        settings.add(dateSetting);
        settings.add(timeSetting);
        settings.add(repeatSetting);
        settings.add(categorySetting);
        settings.add(calenderSetting);
        settings.add(notificationSetting);
        settings.add(durationSetting);
        settings.add(prioritizedSetting);
        settings.add(templateSetting);
        settings.add(tagsSetting);
        settings.add(colourSetting);
        return settings;
    }
    public Mental getMental() {
        log("ItemSessionViewModel.getMental()");
        return currentItem.getMental();
    }
    public void pauseTimer(){
        log("ItemSessionViewModel.pauseTimer()");
        if( kronos == null){
            return;
        }
        mutableTimerState.setValue(Kronos.State.PAUSED);
        kronos.pause();
    }
    public void resumeTimer(){
        log("ItemSessionViewModel.resumeTimer()");
        if(kronos == null){
            log("WARNING, trying to resume a null Timer");
            return;
        }
        mutableTimerState.setValue(Kronos.State.RUNNING);
        kronos.resume();
    }
    public void setElapsedDuration(long secs){
        log("ItemSessionViewModel.setElapsedDuration(long)", secs);
        if( kronos != null) {
            kronos.setElapsedTime(secs);
        }
    }
    public void stopTimer(){
        log("ItemSessionViewModel.stopTimer()");
        if( kronos == null){
            log("WARNING, stopTimer() called with null kronos");
        }

        kronos.stop();
    }
    public boolean update(Item item, Context context){
        log("ItemSessionViewMode.update(Item)", item.getHeading());
        int rowsAffected = ItemsWorker.update(item, context);
        if( rowsAffected != 1){
            mutableError.setValue("ERROR, updating item");
        }else{
            ItemsWorker.touchParents(item, context);
        }
        return rowsAffected == 1;
    }
    public void update(ItemSetting itemSetting, Item item, Context context){
        switch (itemSetting.getKey()) {
            case DATE:
                itemSetting.setValue(item.getTargetDate().toString());
                break;
        }
    }

    public void updateAnxiety(int anxiety, Context context){
        log("...updateAnxiety(int)", anxiety);
        currentItem.setAnxiety(anxiety - MENTAL_OFFSET);
        update(currentItem, context);
    }

    /**
     * updates currentItems energy field and saves it
     * @param energy 0 - 10
     * @param context, context context
     */
    public void updateEnergy(int  energy, Context context) {
        log("ItemSessionViewModel.updateEnergy(int)", energy);
        currentItem.setEnergy(energy - MENTAL_OFFSET);
        update(currentItem, context);
    }

    /**
     *
     * @param mood 0 - 10
     * @param context, context context
     */
    public void updateMood(int mood, Context context){
        log("...updateMood()");
        currentItem.setMood(mood - MENTAL_OFFSET);
        update(currentItem, context);
    }

    /**
     *
     * @param stress, 0 - 10
     * @param context, context
     */
    public void updateStress(int stress, Context context){
        log("...updateStress(int)", stress);
        currentItem.setStress(stress - MENTAL_OFFSET);
        update(currentItem, context);
    }
    public void setIsEvent(boolean checked, Context context) {
        log("...setIsEvent(boolean()", checked);
        currentItem.setIsCalenderItem(checked);
        update(currentItem, context);
    }
    public void setCategory(String category, Context context) {
        log("...setCategory(String, Context)");
        currentItem.setCategory(category);
        update(currentItem, context);
    }

    public void setTargetTime(LocalTime targetTime, Context context) {
        log("...setTargetTime(LocalTime)", targetTime);
        currentItem.setTargetTime(targetTime);
        update(currentItem, context);
    }

    public void setDuration(Duration duration, Context context) {
        log("...setDuration(Duration, Context)");
        currentItem.setDuration(duration.getSeconds());
        mutableDuration.setValue(duration.getSeconds());
        update(currentItem, context);
    }

    public void setDone(boolean checked) {
        //currentItem.setDone();
    }

    public void startTimer(){
        log("ItemSessionViewModel.startTimer()");
        currentTimerItem = currentItem;
        kronos = Kronos.getInstance(secs -> {
            if(VERBOSE )log("Kronos.onTimerTick(long)", secs);
            mutableDuration.setValue(secs);
        });
        mutableTimerState.setValue(Kronos.State.RUNNING);
        kronos.setElapsedTime(currentTimerItem.getDuration());
        kronos.start(currentTimerItem.getID());

    }

    public void setIsTemplate(boolean isTemplate, Context context) {
        log("ItemSessionViewModel.setIsTemplate(boolean, Context)");
        currentItem.setIsTemplate(isTemplate);
        update(currentItem, context);
    }
    public Item getItem() {
        log("ItemSessionViewModel.getItem()");
        return currentItem;
    }

    public boolean itemHasRepeat() {
        log("ItemSessionViewModel.itemHasRepeat(Context)");
        return currentItem.getRepeatID() > 0;
    }

    public void setIsPrioritized(boolean checked, Context context) {
        currentItem.setPriority(checked ? 1:0);
        update(currentItem, context);
    }

    public LiveData<Kronos.State> getTimerState() {
        log("ItemSessionViewModel.getTimerState()");
        if(kronos == null){
            mutableTimerState.setValue(Kronos.State.PENDING);
        }
        return mutableTimerState;
    }


    public void updateNotification(Context context) {
        log("ItemSessionViewModel.updateNotification(Context)");
    }
}

File: app/src/main/java/se/curtrune/lucy/screens/item_editor/ItemSessionViewModel.kt
package se.curtrune.lucy.screens.item_editor

import android.content.Context
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import se.curtrune.lucy.app.LucindaApplication
import se.curtrune.lucy.R
import se.curtrune.lucy.classes.item.Item
import se.curtrune.lucy.classes.Mental
import se.curtrune.lucy.item_settings.CheckBoxSetting
import se.curtrune.lucy.item_settings.ItemSetting
import se.curtrune.lucy.item_settings.KeyValueSetting
import se.curtrune.lucy.services.TimerService
import se.curtrune.lucy.util.DateTImeConverter
import se.curtrune.lucy.util.Logger
import se.curtrune.lucy.features.notifications.NotificationsWorker

class ItemSessionViewModel : ViewModel() {
    private var elapsedTime = 0L
    private val repository = LucindaApplication.appModule.repository
    private val timeModule = LucindaApplication.appModule.timeModule
    private val mutableTimerState = MutableLiveData<TimerState>()
    private val mutableDuration = MutableLiveData<Long>()
    private val mutableError = MutableLiveData<String>()
    private val mutableCurrentItem = MutableLiveData<Item?>()
    private var currentItem: Item? = null

    init{
        mutableTimerState.value = TimerState.PENDING
    }
    fun addTags(tags: String?, context: Context?) {
        currentItem!!.tags = tags
        //update(currentItem!!, context)
    }

    fun cancelNotification(context: Context?) {
        Logger.log("ItemSessionViewModel.cancelNotification(Context)")
        NotificationsWorker.cancelNotification(currentItem, context)
        currentItem!!.notification = null
        repository.update(currentItem!!)
    }

    private fun cancelTimer() {
        Logger.log("ItemSessionViewModel.cancelTimer()")
        timeModule.cancelTimer()
        mutableDuration.value = 0L
        mutableTimerState.value = TimerState.PENDING
    }

    fun init(item: Item, context: Context?) {
        Logger.log("ItemSessionViewModel.init(Item)")
        this.currentItem = item
        if (currentItem!!.repeatID > 0) {
            val repeat =    repository.selectRepeat(currentItem!!.repeatID)
            currentItem!!.setRepeat(repeat)
        }
        mutableCurrentItem.value = currentItem
        mutableDuration.value = currentItem!!.duration
    }

    fun getCurrentItem(): MutableLiveData<Item?> {
        return mutableCurrentItem
    }

    val duration: LiveData<Long>
        get() = mutableDuration

    val error: LiveData<String>
        get() = mutableError
    val stress: Int
        get() = currentItem!!.stress
    val anxiety: Int
        get() = currentItem!!.anxiety
    val mood: Int
        get() = currentItem!!.mood
    val energy: Int
        get() = currentItem!!.energy

    enum class TimerState{
        PENDING, RUNNING, PAUSED,
    }

    fun getItemSettings(item: Item, context: Context): List<ItemSetting> {
        checkNotNull(item)
        Logger.log("ItemEditorViewModel.getItemSettings(Item)")
        val settings: MutableList<ItemSetting> = ArrayList()
        val dateSetting = KeyValueSetting(
            context.getString(R.string.date),
            item.targetDate.toString(),
            ItemSetting.Key.DATE
        )
        val timeSetting = KeyValueSetting(
            context.getString(R.string.time),
            item.targetTime.toString(),
            ItemSetting.Key.TIME
        )
        var repeatValue = ""
        if (item.hasRepeat()) {
            repeatValue = item.repeat.toString()
        }
        val repeatSetting =
            KeyValueSetting(context.getString(R.string.repeat), repeatValue, ItemSetting.Key.REPEAT)
        val colourSetting = KeyValueSetting(
            context.getString(R.string.color),
            item.color.toString(),
            ItemSetting.Key.COLOR
        )
        val templateSetting = CheckBoxSetting(
            context.getString(R.string.is_template),
            item.isTemplate,
            ItemSetting.Key.TEMPLATE
        )
        val prioritizedSetting = CheckBoxSetting(
            context.getString(R.string.is_prioritized),
            item.isPrioritized,
            ItemSetting.Key.PRIORITIZED
        )
        val durationSetting = KeyValueSetting(
            context.getString(R.string.duration),
            DateTImeConverter.formatSeconds(item.duration),
            ItemSetting.Key.DURATION
        )
        var notificationValue = ""

        if (item.hasNotification()) {
            notificationValue = item.notification.toString()
        }
        val notificationSetting =
            KeyValueSetting("notification", notificationValue, ItemSetting.Key.NOTIFICATION)
        val tagsSetting =
            KeyValueSetting(context.getString(R.string.tags), item.tags, ItemSetting.Key.TAGS)
        val calenderSetting = CheckBoxSetting(
            context.getString(R.string.is_calender),
            item.isCalenderItem,
            ItemSetting.Key.IS_CALENDAR_ITEM
        )
        val categorySetting = KeyValueSetting(
            context.getString(R.string.category),
            item.category,
            ItemSetting.Key.CATEGORY
        )

        settings.add(dateSetting)
        settings.add(timeSetting)
        settings.add(repeatSetting)
        settings.add(categorySetting)
        settings.add(calenderSetting)
        settings.add(notificationSetting)
        settings.add(durationSetting)
        settings.add(prioritizedSetting)
        settings.add(templateSetting)
        settings.add(tagsSetting)
        settings.add(colourSetting)
        return settings
    }
    val item: Item?
        get() {
            Logger.log("ItemSessionViewModel.getItem()")
            return currentItem
        }

    fun itemHasRepeat(): Boolean {
        Logger.log("ItemSessionViewModel.itemHasRepeat(Context)")
        return currentItem!!.repeatID > 0
    }

    val mental: Mental
        get() {
            Logger.log("ItemSessionViewModel.getMental()")
            return currentItem!!.mental
        }
    infix fun onEvent(event: ItemEvent){
        when(event){
            is ItemEvent.Update -> { update(event.item) }
            is ItemEvent.CancelTimer -> {cancelTimer()}
            is ItemEvent.PauseTimer -> {pauseTimer()}
            is ItemEvent.StartTimer -> {startTimer()}
            is ItemEvent.ResumeTimer -> {resumeTimer()}
            is ItemEvent.Delete -> { println("delete not implemented at this stage")}
            is ItemEvent.GetChildren -> {}
            is ItemEvent.GetItem -> {}
            is ItemEvent.GetChildrenType -> {}
            is ItemEvent.Edit -> {}
            is ItemEvent.InsertItem -> {}
            is ItemEvent.ShowAddItemDialog -> {}
            is ItemEvent.InsertChild -> {
                insertChild(event.item)
            }
        }
    }
    private fun insertChild(item: Item){
        println("ItemSessionViewModel.insertChild(${item.heading})")
        currentItem?.let { repository.insertChild(it, item) }

    }

    private fun pauseTimer() {
        println("ItemSessionViewModel.pauseTimer()")
        timeModule.pauseTimer()
        mutableTimerState.value = TimerState.PAUSED
    }

    private fun resumeTimer() {
        println("ItemSessionViewModel.resumeTimer()")
        timeModule.resumeTimer()
        mutableTimerState.value = TimerState.RUNNING
        //timerRunning = true
        //runTimer()
    }

    fun setElapsedDuration(secs: Long) {
        println("ItemSessionViewModel.setElapsedDuration(${DateTImeConverter.formatSeconds(secs)})")
        elapsedTime = secs
    }

    private fun startTimer() {
        Logger.log("ItemSessionViewModel.startTimer()")
        timeModule.startTimer(0)
        mutableTimerState.value = TimerState.RUNNING
        TimerService.currentDuration.observeForever { seconds->
            println("observing forever and ever $seconds")
            mutableDuration.value = seconds
        }

/*        currentTimerItem = currentItem
        timerRunning = true
        if(mutableTimerState.value == TimerState.PENDING){
            elapsedTime = 0
        }
        mutableTimerState.value = TimerState.RUNNING
        runTimer()*/
    }

    private fun update(item: Item): Boolean {
        Logger.log("ItemSessionViewModel.update(Item)", item.heading)
        val rowsAffected = repository.update(item)
        if (rowsAffected != 1) {
            mutableError.setValue("ERROR, updating item")
        } else {
            //ItemsWorker.touchParents(item, context)
            repository.touchParents(item)
        }
        return rowsAffected == 1
    }

    fun update(itemSetting: ItemSetting, item: Item, context: Context?) = when (itemSetting.key) {
        ItemSetting.Key.DATE -> {
            itemSetting.value =
                item.targetDate.toString()
        }

        ItemSetting.Key.TIME -> {
            TODO()
        }
        ItemSetting.Key.DURATION -> {
            TODO()
        }
        ItemSetting.Key.COLOR -> {
            TODO()
        }
        ItemSetting.Key.REPEAT -> {
            TODO()
        }
        ItemSetting.Key.NOTIFICATION -> {
            TODO()
        }
        ItemSetting.Key.CATEGORY -> {
            TODO()
        }
        ItemSetting.Key.TAGS -> {
            TODO()
        }
        ItemSetting.Key.APPOINTMENT -> {
            TODO()
        }
        ItemSetting.Key.DONE -> {
            TODO()
        }
        ItemSetting.Key.TEMPLATE -> {
            TODO()
        }
        ItemSetting.Key.PRIORITIZED -> {
            TODO()
        }
        ItemSetting.Key.IS_CALENDAR_ITEM -> {
            TODO()
        }
        else ->{
            println("ELSE ERROR, itemSessionViewModel")
        }
    }

    val timerState: LiveData<TimerState>
        get() {
            println("ItemSessionViewModel.getTimerState()")
            return mutableTimerState
        }


    fun updateNotification(context: Context?) {
        Logger.log("ItemSessionViewModel.updateNotification(Context)")
    }

    companion object {
        private const val MENTAL_OFFSET = 5
        var VERBOSE: Boolean = false
    }
}
